### ISSUE TITLE
Summing an empty tuple with `sum`

### ISSUE THREAD

**Issue by phipsgabler** [JuliaLang/julia#36055]
On Julia 1.3.1.  I wanted to sum over some varargs, and interestingly `sum` fails in two ways:

```julia
julia> sum(())
ERROR: MethodError: reduce_empty(::typeof(Base.add_sum), ::Type{Union{}}) is ambiguous. Candidates:
  reduce_empty(::typeof(Base.add_sum), ::Type{T}) where T<:Union{Int16, Int32, Int8} in Base at reduce.jl:231
  reduce_empty(::typeof(Base.add_sum), ::Type{T}) where T<:Union{UInt16, UInt32, UInt8} in Base at reduce.jl:232
Possible fix, define
  reduce_empty(::typeof(Base.add_sum), ::Type{Union{}})
Stacktrace:
 [1] mapreduce_empty(::typeof(identity), ::Function, ::Type) at ./reduce.jl:247
 [2] mapreduce_empty_iter(::Function, ::Function, ::Tuple{}, ::Base.HasEltype) at ./reduce.jl:254
 [3] mapfoldl_impl at ./tuple.jl:200 [inlined]
 [4] #mapfoldl#186 at ./reduce.jl:72 [inlined]
 [5] mapfoldl at ./reduce.jl:72 [inlined]
 [6] #mapreduce#194 at ./reduce.jl:200 [inlined]
 [7] mapreduce(::Function, ::Function, ::Tuple{}) at ./reduce.jl:200
 [8] sum(::Function, ::Tuple{}) at ./reduce.jl:395
 [9] sum(::Tuple{}) at ./reduce.jl:412
 [10] top-level scope at REPL[85]:1
 [11] eval(::Module, ::Any) at ./boot.jl:330
 [12] eval_user_input(::Any, ::REPL.REPLBackend) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:86
 [13] run_backend(::REPL.REPLBackend) at /home/philipp/.julia/packages/Revise/AMRie/src/Revise.jl:1023
 [14] top-level scope at REPL[2]:0

julia> sum((); init=0)
ERROR: MethodError: no method matching sum(::Tuple{}; init=0)
Closest candidates are:
  sum(::Any) at reduce.jl:412 got unsupported keyword argument "init"
  sum(::Any, ::StaticArrays.StaticArray{#s160,T,N} where N where #s160<:Tuple; dims) where T at /home/philipp/.julia/packages/StaticArrays/mlIi1/src/mapreduce.jl:251 got unsupported keyword argument "init"
  sum(::Any, ::AbstractArray; dims) at reducedim.jl:653 got unsupported keyword argument "init"
  ...
Stacktrace:
 [1] top-level scope at REPL[86]:1
 [2] eval(::Module, ::Any) at ./boot.jl:330
 [3] eval_user_input(::Any, ::REPL.REPLBackend) at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.3/REPL/src/REPL.jl:86
 [4] run_backend(::REPL.REPLBackend) at /home/philipp/.julia/packages/Revise/AMRie/src/Revise.jl:1023
 [5] top-level scope at REPL[2]:0
```

OK, the first one is reasonable, although I wouldn't have expected an ambiguity error.  But the second one really should work, IMO.

I'll stick to `reduce(+, (), init=0)` for now, since that works.

**Comment by KristofferC** [comment]
Fixed in 1.5 with

```
julia> sum(())
ERROR: ArgumentError: reducing over an empty collection is not allowed
```

**Comment by phipsgabler** [comment]
> For the second, I don't think `sum` takes an init argument and you would have to use `reduce`.

Oh, right, I just expected it to (Python does have one, for comparison).  And it was shadowed by the other in my actual function.
